plugins/router/template/router.go

```diff
@@ -252,6 +280,15 @@ func (r *templateRouter) AddEndpoints(id string, endpoints []Endpoint) {
 	r.state[id] = frontend
 }
 
+// cleanUpServiceAliasConfig performs any necessary steps to clean up a service alias config before deleting it from
+// the router.  Right now the only clean up step is to remove any of the certificates on disk.
+func (r *templateRouter) cleanUpServiceAliasConfig(cfg *ServiceAliasConfig) {
+	err := r.certManager.deleteCertificatesForConfig(cfg)
+	if err != nil {
+		glog.Errorf("Error deleting certificates for route %s, the route will still be deleted but files may remain in the container: %v", cfg.Host, err)
```

Comment:
Thought about this one today and I think I have a sane design in my head on how to move forward.  Here is the breakdown of the issues:

1.  delete certs when deleting a route - covered by this PR
2.  overwrite certs for routes - covered by this PR, they're written as ns-routename.pem and an update to an existing route would overwrite
3.  don't write files that you don't have to - not covered.  I think this is small enough to include in this PR.  The cert manager can calc a checksum and add it to the internal certificate object.  If a request to write a cert that has a checksum on it is found it can be ignore
4.  create a sync loop to validate cert checksums in memory against what's on disk - not covered.  Probably too big to be in this PR

The Handle* methods currently block each other when manipulating state and performing a Commit which also writes all the certscerts.  I would like to refactor it two ways.  

1.  Remove the Commit from the Handle* path - this coincides with my comment about state transitions here https://github.com/openshift/origin/pull/1906#discussion_r29071514 and have it operate on an update channel like the kubelet
2.  Remove writing certs from the Commit path - the router can still evaluate the templates but should only include "ready" state routes which means the certs have been written by a 
3.  Introduce a new cert sync loop that writes certs and a cert checksum validation loop that checks checksums against files on disk
4.  Introduce a mutex on the ServiceAliasConfig that would be used by the cert manager when writing or deleting certificates (and by proxy the cert sync loop and delete operation of HandleRoute) and used by the checksum loop to stop changes to the files for that particular route while it's checking it.
5.  Change the internal state map to use pointers, the checksum sync loop would remove the checksum of any non-matching route which would in turn force a rewrite by the cert sync loop

I think this will minimize the blocking of Handle* methods and still provide the stability the checksum loop requires.  I will want to sketch it all out to make sure I'm not missing some race or contention issue but I think it will work.

/cc @pmorie since we discussed it a bit in chat